File: src/main/java/io/vlingo/gradle/ActorProxyGeneratorTask.java
package io.vlingo.gradle;

import io.vlingo.actors.Properties;
import io.vlingo.actors.ProxyGenerator;

import org.gradle.api.DefaultTask;
import org.gradle.api.file.ConfigurableFileCollection;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.provider.SetProperty;
import org.gradle.api.tasks.Classpath;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.OutputDirectory;
import org.gradle.api.tasks.TaskAction;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import java.util.Set;


// TODO isolate using the Gradle Worker API
public class ActorProxyGeneratorTask extends DefaultTask {

    @Classpath
    public final ConfigurableFileCollection classesDirs = getProject().files();

    @Input
    public final SetProperty<String> actorProtocols = getProject().getObjects().setProperty(String.class).empty();

    @OutputDirectory
    public final DirectoryProperty destinationDirectory = getProject().getObjects().directoryProperty();

    @TaskAction
    public void generateActorProxies() throws Exception {
        Set<String> protocols = actorProtocols.get();
        if (!protocols.isEmpty()) {

            // TODO this won't scale with multiple jvm languages
            File classesDir = classesDirs.getSingleFile();
            assert classesDir.isDirectory();

            File destinationDir = destinationDirectory.getAsFile().get();
            if (destinationDir.exists()) {
                Files.walk(destinationDir.toPath()).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
            }
            destinationDir.mkdirs();

            // TODO this won't play well with concurrent task execution
            Properties.properties.setProperty("proxy.generated.classes.main", classesDir.getCanonicalPath() + "/");
            Properties.properties.setProperty("proxy.generated.sources.main", destinationDir.getCanonicalPath() + "/");

            try (ProxyGenerator generator = ProxyGenerator.forMain(true)) {
                for (String protocol : protocols) {
                    generator.generateFor(protocol);
                }
            }
        }
    }
}

File: src/main/kotlin/io/vlingo/gradle/ActorProxyGeneratorTask.kt
package io.vlingo.gradle

import org.gradle.api.DefaultTask
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.tasks.CacheableTask
import org.gradle.api.tasks.Classpath
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity
import org.gradle.api.tasks.TaskAction
import org.gradle.workers.IsolationMode
import org.gradle.workers.WorkerExecutor

import org.gradle.kotlin.dsl.*

import javax.inject.Inject


@CacheableTask
open class ActorProxyGeneratorTask @Inject constructor(

    private
    val workerExecutor: WorkerExecutor

) : DefaultTask() {

    @Classpath
    val classpath = project.files()

    @Input
    val actorProtocols = project.objects.setProperty(String::class.java).also { it.set(emptySet()) }

    @OutputDirectory
    @PathSensitive(PathSensitivity.RELATIVE)
    val destinationDirectory = directoryPropertyCompatible()

    @TaskAction
    @Suppress("unused")
    fun generateActorProxies() {
        val destinationDir = destinationDirectory.asFile.get().apply {
            deleteRecursively()
            mkdirs()
        }
        val protocols = actorProtocols.get()
        if (protocols.isNotEmpty()) {

            // TODO:performance let ProxyGenerator accept a ClassLoader, remove classPath from the spec and use current loader in the work
            val spec = ActorProxyGeneratorSpec(
                classpath.files,
                protocols,
                destinationDir
            )

            workerExecutor.submit(ActorProxyGeneratorWork::class) {
                isolationMode = IsolationMode.CLASSLOADER
                classpath = this@ActorProxyGeneratorTask.classpath.files
                params(spec)
            }
        } else {
            didWork = false
        }
    }

    private
    fun directoryPropertyCompatible(): DirectoryProperty =
        if (hasObjectFactoryPropertyFactories) project.objects.directoryProperty()
        else newOutputDirectory()
}
